File: src/main/java/org/blacksun/network/Network.java
package org.blacksun.network;

import guru.nidi.graphviz.attribute.Color;
import guru.nidi.graphviz.attribute.Label;
import guru.nidi.graphviz.model.MutableGraph;
import guru.nidi.graphviz.model.MutableNode;
import org.blacksun.graph.algorithms.BFAlgorithmFactory;
import org.blacksun.graph.algorithms.GraphPath;
import org.blacksun.graph.algorithms.PathFindingAlgorithm;
import org.blacksun.graph.algorithms.PathFindingAlgorithmFactory;
import org.blacksun.graph.channel.Channel;
import org.blacksun.graph.node.GraphNode;
import org.blacksun.utils.StringRepresentable;
import org.blacksun.utils.Config;
import org.jetbrains.annotations.NotNull;

import java.util.*;
import java.util.stream.Collectors;

import static guru.nidi.graphviz.model.Factory.*;

public class Network implements StringRepresentable {
    private List<GraphNode> nodes;
    private final Topology topology;
    private PathFindingAlgorithmFactory factory;
    private List<GraphPath> openConnections;

    public Network() {
        this(new Topology() {
            @Override
            public List<GraphNode> createNetwork() {
                return new ArrayList<>();
            }

            @Override
            public GraphNode createNode() {
                return null;
            }
        });
    }

    public Network(@NotNull Topology topology) {
        this(topology, new BFAlgorithmFactory());
    }

    public Network(@NotNull Topology topology, PathFindingAlgorithmFactory factory) {
        this.topology = topology;
        this.factory = factory;
        generateNetwork();
    }

    public void generateNetwork() {
        nodes = topology.createNetwork();
        openConnections = new ArrayList<>();
    }

    public List<GraphNode> getNodes() {
        return nodes;
    }

    public List<GraphNode> getNodes(boolean terminal) {
        if (!terminal)
            return nodes;
        return nodes.stream()
                .filter(GraphNode::isTerminal)
                .collect(Collectors.toList());
    }

    public GraphNode getRandomNode() {
        return nodes.get(new Random().nextInt(nodes.size()));
    }

    public GraphNode getRandomNode(boolean terminal) {
        List<GraphNode> nodes = getNodes(terminal);
        return nodes.get(new Random().nextInt(nodes.size()));
    }

    public PathFindingAlgorithmFactory getFactory() {
        return factory;
    }

    public void setFactory(PathFindingAlgorithmFactory factory) {
        this.factory = factory;
    }

    public GraphPath getPath(@NotNull GraphNode from, @NotNull GraphNode to) {
        return getPath(from, to, false);
    }

    public GraphPath getPath(@NotNull GraphNode from, @NotNull GraphNode to, boolean isUsed) {
        if (!exists(from, to))
            return new GraphPath();
        return factory.getAlgorithm(nodes, getChannels(isUsed)).getPath(from, to);
    }

    public List<GraphPath> getPaths(@NotNull GraphNode from) {
        PathFindingAlgorithm algorithm = factory.getAlgorithm(nodes, getChannels());
        return nodes.stream()
                .filter(node -> !node.equals(from))
                .map(node -> algorithm.getPath(from, node))
                .collect(Collectors.toList());
    }

    private List<Channel> getChannels() {
        return nodes.stream()
                .flatMap(node -> node.getConnections().stream())
                .distinct()
                .collect(Collectors.toList());
    }

    private List<Channel> getChannels(boolean isUsed) {
        return nodes.stream()
                .flatMap(node -> node.getConnections().stream())
                .distinct()
                .filter(ch -> ch.isUsed() == isUsed)
                .collect(Collectors.toList());
    }

    public GraphPath createConnection(@NotNull GraphNode from, @NotNull GraphNode to) {
        if (from == to)
            return new GraphPath();
        GraphPath path = getPath(from, to);
        if (path.exists()) {
            path.forEach(ch -> ch.setUsed(true));
            openConnections.add(path);
        }
        return path;
    }

    public void closeConnection(@NotNull GraphPath path) {
        if (openConnections.contains(path)) {
            path.forEach(ch -> ch.setUsed(false));
            openConnections.remove(path);
        }
    }


    public void closeConnection(@NotNull GraphNode from, @NotNull GraphNode to) {
        openConnections.stream()
                .filter(path -> path.ofNodes(from, to))
                .findAny()
                .ifPresent(this::closeConnection);
    }

    public GraphNode addNode() {
        GraphNode node = topology.createNode();
        if (node != null) {
            nodes.add(node);
        }
        return node;
    }

    public void addNode(@NotNull GraphNode node) {
        if (!exists(node)) {
            nodes.add(node);
        }
    }

    public void removeNode(@NotNull GraphNode node) {
        Channel[] connections = node.getConnections().toArray(new Channel[]{});
        for (Channel ch: connections) {
            ch.remove();
        }
        nodes.remove(node);
    }

    public void addConnection(@NotNull GraphNode from, @NotNull GraphNode to, int weight) {
        if (exists(from, to)) {
            from.addConnectedNode(to, weight);
        }
    }

    public void removeConnection(@NotNull GraphNode from, @NotNull GraphNode to) {
        if (exists(from, to)) {
            from.removeConnectedNode(to);
        }
    }

    private boolean exists(GraphNode... nodes) {
        return Arrays.stream(nodes)
                .allMatch(node -> this.nodes.contains(node));
    }

    // TODO(cache?)
    double getAvgOrder() {
        if (nodes.isEmpty())
            return 0;
        double sum = 0;
        for (GraphNode node : nodes) {
            sum += node.getOrder();
        }
        return sum / nodes.size();
    }

    @Override
    public String stringRepresentation() {
        StringBuilder sb = new StringBuilder();
        sb.append("Network[order=")
                .append(getAvgOrder())
                .append("] {\n");
        nodes.forEach(node -> sb.append(node.stringRepresentation()).append(",\n"));
        sb.append("}");
        return sb.toString();
    }

    public MutableGraph toGraph() {
        Config cfg = Config.getConfig();
        HashMap<GraphNode, MutableNode> map = new HashMap<>();
        MutableGraph g = mutGraph("network")
                .setDirected(true);
        nodes.forEach(node -> {
            Color color = cfg.getColor("node");
            if (node.isTerminal())
                color = cfg.getColor("terminal");
            if (node.isSelected())
                color = cfg.getColor("selectedN");
            map.put(node, mutNode(node.toString()).add(color));
        });
        nodes.forEach(node -> {
            MutableNode mNode = map.get(node);
            g.add(mNode);
            node.getConnections().forEach(ch -> {
                MutableNode toNode = map.get(ch.getToNode());
                Color color = cfg.getColor("channel");
                if (ch.isUsed())
                    color = cfg.getColor("connected");
                if (ch.isSelected())
                    color = cfg.getColor("selectedC");
                mNode.addLink(to(toNode).
                        with(Label.markdown(String.valueOf(ch.getWeight())),
                                color));
            });
        });
        return g;
    }

    @Override
    public String toString() {
        return "Network[nodes=" + nodes.size() + ", order=" + getAvgOrder() + "]";
    }

    public void closeAll() {
        getChannels(true).forEach(ch -> ch.setUsed(false));
    }

    public void clear() {
        nodes.clear();
        openConnections.clear();
    }
}

File: src/main/java/org/blacksun/network/Network.kt
package org.blacksun.network

import guru.nidi.graphviz.attribute.Label
import guru.nidi.graphviz.model.Factory
import guru.nidi.graphviz.model.MutableGraph
import guru.nidi.graphviz.model.MutableNode
import org.blacksun.graph.algorithms.BFAlgorithmFactory
import org.blacksun.graph.algorithms.GraphPath
import org.blacksun.graph.algorithms.PathFindingAlgorithmFactory
import org.blacksun.graph.channel.Channel
import org.blacksun.graph.node.GraphNode
import org.blacksun.utils.StringRepresentable
import org.blacksun.utils.Config

import java.util.*

import guru.nidi.graphviz.model.Factory.*
import org.blacksun.graph.channel.SimplexChannelFactory
import org.blacksun.graph.node.SimpleNode

class Network @JvmOverloads constructor(private val topology: Topology = object : Topology {
    override fun createNetwork(): MutableList<GraphNode> {
        return ArrayList()
    }

    override fun createNode(): GraphNode {
        return SimpleNode("Node", SimplexChannelFactory())
    }
}, var factory: PathFindingAlgorithmFactory = BFAlgorithmFactory()) : StringRepresentable {
    private var nodes: MutableList<GraphNode> = ArrayList()
    private var openConnections: MutableList<GraphPath> = ArrayList()

    val randomNode: GraphNode
        get() = nodes[Random().nextInt(nodes.size)]

    private val channels: List<Channel>
        get() = nodes
                .flatMap { node -> node.connections }
                .distinct()

    // TODO(cache?)
    internal val avgOrder: Double
        get() {
            if (nodes.isEmpty())
                return .0
            var sum = .0
            for (node in nodes) {
                sum += node.order.toDouble()
            }
            return sum / nodes.size
        }

    init {
        generateNetwork()
    }

    fun generateNetwork() {
        nodes = topology.createNetwork()
        openConnections = ArrayList()
    }

    fun getNodes(terminal: Boolean): List<GraphNode>? {
        return if (!terminal)
            nodes
        else
            nodes.filter(GraphNode::isTerminal)
    }

    fun getRandomNode(terminal: Boolean): GraphNode {
        val nodes = getNodes(terminal)
        return nodes!![Random().nextInt(nodes.size)]
    }

    @JvmOverloads
    fun getPath(from: GraphNode, to: GraphNode, isUsed: Boolean = false): GraphPath {
        return if (!exists(from, to)) GraphPath() else factory.getAlgorithm(nodes, getChannels(isUsed)).getPath(from, to)
    }

    fun getPaths(from: GraphNode): List<GraphPath> {
        val algorithm = factory.getAlgorithm(nodes, channels)
        return nodes
                .filter {  it != from }
                .map {algorithm.getPath(from, it) }
    }

    private fun getChannels(isUsed: Boolean): List<Channel> {
        return nodes
                .flatMap(GraphNode::connections)
                .distinct()
                .filter { it.isUsed == isUsed }
    }

    fun createConnection(from: GraphNode, to: GraphNode): GraphPath {
        if (from === to)
            return GraphPath()
        val path = getPath(from, to)
        if (path.exists()) {
            path.forEach { it.isUsed = true }
            openConnections.add(path)
        }
        return path
    }

    fun closeConnection(path: GraphPath) {
        if (openConnections.contains(path)) {
            path.forEach { ch -> ch.isUsed = false }
            openConnections.remove(path)
        }
    }


    fun closeConnection(from: GraphNode, to: GraphNode) {
        openConnections.stream()
                .filter { path -> path.ofNodes(from, to) }
                .findAny()
                .ifPresent { this.closeConnection(it) }
    }

    fun addNode(): GraphNode {
        val node = topology.createNode()
        nodes.add(node)
        return node
    }

    fun addNode(node: GraphNode) {
        if (!exists(node)) {
            nodes.add(node)
        }
    }

    fun removeNode(node: GraphNode) {
        val connections = node.connections.toTypedArray()
        for (ch in connections) {
            ch.remove()
        }
        nodes.remove(node)
    }

    fun addConnection(from: GraphNode, to: GraphNode, weight: Int) {
        if (exists(from, to)) {
            from.addConnectedNode(to, weight)
        }
    }

    fun removeConnection(from: GraphNode, to: GraphNode) {
        if (exists(from, to)) {
            from.removeConnectedNode(to)
        }
    }

    private fun exists(vararg nodes: GraphNode): Boolean {
        return nodes.all { it in nodes }
    }

    override fun stringRepresentation(): String {
        val sb = StringBuilder()
        sb.append("Network[order=")
                .append(avgOrder)
                .append("] {\n")
        nodes.forEach { sb.append(it.stringRepresentation()).append(",\n") }
        sb.append("}")
        return sb.toString()
    }

    fun toGraph(): MutableGraph {
        val cfg = Config.config
        val map = HashMap<GraphNode, MutableNode>()
        val g = mutGraph("network")
                .setDirected(true)
        nodes.forEach { node ->
            var color = cfg.getColor("node")
            if (node.isTerminal)
                color = cfg.getColor("terminal")
            if (node.isSelected)
                color = cfg.getColor("selectedN")
            map[node] = mutNode(node.toString()).add(color)
        }
        nodes.forEach { node ->
            val mNode = map[node]!!
            g.add(mNode)
            node.connections.forEach { ch ->
                val toNode = map[ch.toNode]!!
                var color = cfg.getColor("channel")
                if (ch.isUsed)
                    color = cfg.getColor("connected")
                if (ch.isSelected)
                    color = cfg.getColor("selectedC")
                mNode.addLink(Factory.to(toNode).with(Label.markdown(ch.weight.toString()),
                        color))
            }
        }
        return g
    }

    override fun toString(): String {
        return "Network[nodes=" + nodes.size + ", order=" + avgOrder + "]"
    }

    fun closeAll() {
        getChannels(true).forEach { ch -> ch.isUsed = false }
    }

    fun clear() {
        nodes.clear()
        openConnections.clear()
    }
}
